File: backend/src/services/user.service.ts
import { RefreshToken } from "../models/refresh-token.model.js";
import { Order } from "../models/order.model.js";
import { User, type IUser, type IUserAddress } from "../models/user.model.js";
import type {
  ChangePasswordInput,
  UserAddressInput,
  UserAddressListResponse,
  UserAddressResponse,
  UserAddressUpdateInput,
  UserMessageResponse,
  UserProfileResponse,
  UserProfileUpdateInput,
} from "../types/user.types.js";

export class UserServiceError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "UserServiceError";
    this.statusCode = statusCode;
  }
}

function toAddressResponse(address: IUserAddress): UserAddressResponse {
  return {
    _id: address._id?.toString() ?? "",
    ...(address.label === undefined ? {} : { label: address.label }),
    address: address.address,
    isDefault: address.isDefault,
  };
}

function toProfileResponse(user: IUser): UserProfileResponse {
  return {
    id: user._id.toString(),
    email: user.email,
    name: user.name,
    ...(user.phone === undefined ? {} : { phone: user.phone }),
    ...(user.avatar === undefined ? {} : { avatar: user.avatar }),
    role: user.role,
    status: user.status,
    isActive: user.isActive !== false,
    addresses: user.addresses.map(toAddressResponse),
  };
}

async function requireUser(userId: string): Promise<IUser> {
  const user = await User.findById(userId);

  if (user === null || user.status !== "active" || user.isActive === false) {
    throw new UserServiceError("User not found", 404);
  }

  return user;
}

function findAddress(user: IUser, addressId: string): IUserAddress {
  const address = user.addresses.find(
    (item) => item._id !== undefined && item._id.toString() === addressId,
  );

  if (address === undefined) {
    throw new UserServiceError("Address not found", 404);
  }

  return address;
}

function clearDefaultAddresses(user: IUser): void {
  for (const address of user.addresses) {
    address.isDefault = false;
  }
}

export async function getUserProfile(
  userId: string,
): Promise<UserProfileResponse> {
  return toProfileResponse(await requireUser(userId));
}

export async function updateUserProfile(
  userId: string,
  input: UserProfileUpdateInput,
): Promise<UserProfileResponse> {
  const user = await requireUser(userId);

  if (input.name !== undefined) {
    user.name = input.name;
  }

  if (input.phone !== undefined) {
    user.phone = input.phone;
  }

  if (input.avatar !== undefined) {
    user.avatar = input.avatar;
  }

  await user.save();
  return toProfileResponse(user);
}

export async function changeUserPassword(
  userId: string,
  input: ChangePasswordInput,
): Promise<UserMessageResponse> {
  const user = await User.findById(userId).select("+password");

  if (user === null || user.status !== "active" || user.isActive === false) {
    throw new UserServiceError("User not found", 404);
  }

  if ((await user.matchPassword(input.currentPassword)) === false) {
    throw new UserServiceError("Current password is incorrect", 401);
  }

  user.password = input.newPassword;
  await user.save();
  await RefreshToken.deleteMany({ userId: user._id });

  return {
    message: "Password changed",
  };
}

export async function addUserAddress(
  userId: string,
  input: UserAddressInput,
): Promise<UserProfileResponse> {
  const user = await requireUser(userId);
  const isDefault = input.isDefault === true || user.addresses.length === 0;

  if (isDefault === true) {
    clearDefaultAddresses(user);
  }

  user.addresses.push({
    ...(input.label === undefined ? {} : { label: input.label }),
    address: input.address,
    isDefault,
  });

  await user.save();
  return toProfileResponse(user);
}

export async function listUserAddresses(
  userId: string,
): Promise<UserAddressListResponse> {
  const user = await requireUser(userId);

  return {
    records: user.addresses.map(toAddressResponse),
  };
}

export async function setDefaultUserAddress(
  userId: string,
  addressId: string,
): Promise<UserProfileResponse> {
  const user = await requireUser(userId);
  const address = findAddress(user, addressId);

  clearDefaultAddresses(user);
  address.isDefault = true;
  await user.save();
  return toProfileResponse(user);
}

export async function updateUserAddress(
  userId: string,
  addressId: string,
  input: UserAddressUpdateInput,
): Promise<UserProfileResponse> {
  const user = await requireUser(userId);
  const address = findAddress(user, addressId);

  if (input.label !== undefined) {
    address.label = input.label;
  }

  if (input.address !== undefined) {
    address.address = input.address;
  }

  if (input.isDefault === true) {
    clearDefaultAddresses(user);
    address.isDefault = true;
  } else if (input.isDefault === false) {
    address.isDefault = false;
  }

  await user.save();
  return toProfileResponse(user);
}

export async function deleteUserAddress(
  userId: string,
  addressId: string,
): Promise<UserProfileResponse> {
  const user = await requireUser(userId);
  const addressIndex = user.addresses.findIndex(
    (address) =>
      address._id !== undefined && address._id.toString() === addressId,
  );

  if (addressIndex === -1) {
    throw new UserServiceError("Address not found", 404);
  }

  const [removedAddress] = user.addresses.splice(addressIndex, 1);

  if (
    removedAddress?.isDefault === true &&
    user.addresses.length > 0 &&
    user.addresses.some((address) => address.isDefault === true) === false
  ) {
    const firstAddress = user.addresses[0];

    if (firstAddress !== undefined) {
      firstAddress.isDefault = true;
    }
  }

  await user.save();
  return toProfileResponse(user);
}

export async function deactivateCurrentUser(
  userId: string,
): Promise<UserMessageResponse> {
  const blockingOrder = await Order.exists({
    userId,
    status: { $in: ["pending"] },
  });

  if (blockingOrder !== null) {
    throw new UserServiceError(
      "Pending orders must finish before deleting the account",
      409,
    );
  }

  const user = await requireUser(userId);
  user.isActive = false;
  user.status = "banned";
  await user.save();
  await RefreshToken.deleteMany({ userId: user._id });

  return {
    message: "Tai khoan da duoc xoa (vo hieu hoa)",
  };
}

File: backend-springboot/src/main/java/com/season/app/service/UserService.java
package com.season.app.service;

import com.season.app.exception.UserServiceException;
import com.season.app.model.User;
import com.season.app.model.UserAddress;
import com.season.app.repository.UserRepository;
// import com.season.app.repository.OrderRepository;
// import com.season.app.repository.RefreshTokenRepository;
import com.season.app.dto.*; // Import các class DTO (sẽ hướng dẫn tạo sau)

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;

    // TODO: Sẽ cần tạo OrderRepository và RefreshTokenRepository sau này
    // @Autowired
    // private OrderRepository orderRepository;
    
    // @Autowired
    // private RefreshTokenRepository refreshTokenRepository;

    /**
     * Hàm Helper: Tìm user và kiểm tra trạng thái hoạt động
     * Tương đương: requireUser(userId) trong file TS
     */
    private User requireUser(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new UserServiceException(HttpStatus.NOT_FOUND, "User not found"));

        if (!"active".equals(user.getStatus()) || Boolean.FALSE.equals(user.getIsActive())) {
            throw new UserServiceException(HttpStatus.NOT_FOUND, "User not found");
        }
        return user;
    }

    /**
     * Hàm Helper: Tìm địa chỉ cụ thể của một user
     * Tương đương: findAddress(user, addressId)
     */
    private UserAddress findAddress(User user, String addressId) {
        return user.getAddresses().stream()
                .filter(a -> a.getId().equals(addressId))
                .findFirst()
                .orElseThrow(() -> new UserServiceException(HttpStatus.NOT_FOUND, "Address not found"));
    }

    /**
     * Hàm Helper: Xoá cờ mặc định của tất cả địa chỉ
     * Tương đương: clearDefaultAddresses(user)
     */
    private void clearDefaultAddresses(User user) {
        user.getAddresses().forEach(a -> a.setIsDefault(false));
    }

    // ==========================================
    // CÁC HÀM NGHIỆP VỤ CHÍNH (PUBLIC METHODS)
    // ==========================================

    public UserProfileResponse getUserProfile(String userId) {
        User user = requireUser(userId);
        return new UserProfileResponse(user); // Map từ Entity sang DTO để trả về
    }

    public UserProfileResponse updateUserProfile(String userId, UserProfileUpdateInput input) {
        User user = requireUser(userId);

        if (input.getName() != null) user.setName(input.getName());
        if (input.getPhone() != null) user.setPhone(input.getPhone());
        if (input.getAvatar() != null) user.setAvatar(input.getAvatar());

        userRepository.save(user); // Lưu vào MongoDB
        return new UserProfileResponse(user);
    }

    public UserMessageResponse changeUserPassword(String userId, ChangePasswordInput input) {
        User user = requireUser(userId);

        // Kiểm tra mật khẩu cũ (Tương đương user.matchPassword)
        if (!passwordEncoder.matches(input.getCurrentPassword(), user.getPassword())) {
            throw new UserServiceException(HttpStatus.UNAUTHORIZED, "Current password is incorrect");
        }

        // Mã hoá mật khẩu mới và lưu lại
        user.setPassword(passwordEncoder.encode(input.getNewPassword()));
        userRepository.save(user);

        // Xoá tất cả refresh token để bắt đăng nhập lại trên các thiết bị
        // refreshTokenRepository.deleteAllByUserId(user.getId());

        return new UserMessageResponse("Password changed");
    }

    public UserProfileResponse addUserAddress(String userId, UserAddressInput input) {
        User user = requireUser(userId);
        
        boolean isDefault = Boolean.TRUE.equals(input.getIsDefault()) || user.getAddresses().isEmpty();

        if (isDefault) {
            clearDefaultAddresses(user);
        }

        UserAddress newAddress = UserAddress.builder()
                .label(input.getLabel())
                .address(input.getAddress())
                .isDefault(isDefault)
                .build();

        user.getAddresses().add(newAddress);
        userRepository.save(user);
        
        return new UserProfileResponse(user);
    }

    public List<UserAddress> listUserAddresses(String userId) {
        User user = requireUser(userId);
        return user.getAddresses();
    }

    public UserProfileResponse setDefaultUserAddress(String userId, String addressId) {
        User user = requireUser(userId);
        UserAddress address = findAddress(user, addressId);

        clearDefaultAddresses(user);
        address.setIsDefault(true);
        
        userRepository.save(user);
        return new UserProfileResponse(user);
    }

    public UserProfileResponse updateUserAddress(String userId, String addressId, UserAddressUpdateInput input) {
        User user = requireUser(userId);
        UserAddress address = findAddress(user, addressId);

        if (input.getLabel() != null) address.setLabel(input.getLabel());
        if (input.getAddress() != null) address.setAddress(input.getAddress());

        if (Boolean.TRUE.equals(input.getIsDefault())) {
            clearDefaultAddresses(user);
            address.setIsDefault(true);
        } else if (Boolean.FALSE.equals(input.getIsDefault())) {
            address.setIsDefault(false);
        }

        userRepository.save(user);
        return new UserProfileResponse(user);
    }

    public UserProfileResponse deleteUserAddress(String userId, String addressId) {
        User user = requireUser(userId);
        
        // Tìm địa chỉ cần xoá
        Optional<UserAddress> addressToRemove = user.getAddresses().stream()
                .filter(a -> a.getId().equals(addressId))
                .findFirst();

        if (addressToRemove.isEmpty()) {
            throw new UserServiceException(HttpStatus.NOT_FOUND, "Address not found");
        }

        UserAddress removedAddress = addressToRemove.get();
        user.getAddresses().remove(removedAddress);

        // Nếu xoá địa chỉ mặc định, tự động gán địa chỉ đầu tiên còn lại thành mặc định
        if (Boolean.TRUE.equals(removedAddress.getIsDefault()) && !user.getAddresses().isEmpty()) {
            boolean hasDefault = user.getAddresses().stream().anyMatch(a -> Boolean.TRUE.equals(a.getIsDefault()));
            if (!hasDefault) {
                user.getAddresses().get(0).setIsDefault(true);
            }
        }

        userRepository.save(user);
        return new UserProfileResponse(user);
    }

    public UserMessageResponse deactivateCurrentUser(String userId) {
        // Kiểm tra xem có order nào đang pending không
        // boolean hasPendingOrder = orderRepository.existsByUserIdAndStatusIn(userId, List.of("pending"));
        // if (hasPendingOrder) {
        //     throw new UserServiceException(HttpStatus.CONFLICT, "Pending orders must finish before deleting the account");
        // }

        User user = requireUser(userId);
        user.setIsActive(false);
        user.setStatus("banned"); // Theo nguyên bản Node.js, status chuyển thành banned
        
        userRepository.save(user);
        
        // refreshTokenRepository.deleteAllByUserId(user.getId());

        return new UserMessageResponse("Tai khoan da duoc xoa (vo hieu hoa)");
    }
}